docs/rfc/rfc-001-storage-engine.rst

```diff
@@ -0,0 +1,179 @@
+platform, that the 0.36 release presents a good opportunity to correct. The
+current storage engine layer provides a great deal of flexibility that is
+difficult for users to leverage or benefit from, while also making it harder
+for Tendermint Core developers to deliver improvements on storage engine. This
+RFC discusses the possible improvements to this layer of the system.
+
+Background
+----------
+
+Tendermint has a very thin common wrapper that makes Tendermint itself
+(largely) agnostic to the data storage layer (within the realm of the popular
+key-value/embedded databases.) This flexibility is not particularly useful:
+the benefits of a specific database engine in the context of tendermint is not
+particularly well understood, and the maintenance burden for multiple backends
+is not commensurate with the benefit provided. Additionally, because the data
+storage layer is handled generically, and most tests run with an in-memory
+framework, it's difficult to take advantage of any higher-level features of a
+database engine.
+
+Ideally, developers within Tendermint will be able to interact with persisted
+data via a single interface that can function, approximately like an object
+store, and this storage interface will be able to accommodate all existing
+persistence workloads (e.g. block storage, local peer management information
+like the "address book", crash-recovery log like the WAL.) In addition to
+providing a more ergonomic interface and new semantics, by selecting a single
+storage engine tendermint can use native durability and atomicity features of
+the storage engine and simplify its own implementations. 
+
+Data Access Patterns
+~~~~~~~~~~~~~~~~~~~~
+
+Tendermint's data access patterns have the following characteristics:
+
+- data size often exceeds memory.
+
+- data is rarely mutated after it's written for most data (e.g. blocks), but
+  small amounts of workding data is persisted by nodes and is frequently
+  mutated (e.g. peer information, validator information.)
+
+- read patterns can be quite random.
+
+- crash resistance and crash recovery, provided by write-ahead-logs (in
+  consensus, and potentially for the mempool) should allow the system to
+  resume work after an unexpected shut down.
+
+Project Goals
+~~~~~~~~~~~~~
+
+As we think about replacing the current persistence layer, we should consider
+the following high level goals: 
+
+- drop dependencies on storage engines that have a CGo dependency.
+
+- encapsulate data format and data storage from higher-level services
+  (e.g. reactors) within tendermint.
+
+- select a storage engine that does not incur any additional operational
+  complexity (e.g. database should be embedded.)
+
+- provide databases semantics with sufficient ACID, snapshots, and
+  transactional support.
+
+Open Questions
+~~~~~~~~~~~~~~
+
+The following questions remain:
+
+- what kind of data-access concurrency does tendermint require?
```

Comment:
This is a good question. Though if we're assuming Tendermint mediates access to storage through (an) interface(s), this may not matter that much—the node can synchronize access.

If we _don't_ have the node mediate, then the choice of possible backing storage gets much narrower. A lot of popular storage libraries allow concurrent access from a single process, but lock between processes. The ones that don't tend to be more DBMS-like (e.g., Postgres) and not easily embedded. Do you have any particular candidates in mind?

Reply:
I think Badger is the storage engine to beat for us, though I think we could definitely build what we need on top of bolt (though bolt is less maintained/more simple.) 

Badger's interface/transactions can be used from multiple threads internally so we don't need to add an extra mutex, though you're right, we can if that's needed.

Reply:
> If blocks are never mutated and are append only do we even need a locking mechanism? We're not going to have a situation of conflicting reads/writes

I presume at some point we will need to prune old data from the log, though. So even if we don't need to worry about concurrent access most of the time, we would presumably need some way to either lock out readers during a pruning stage, or stop-the-world and restart the node?

Reply:
Re: Badger. Another approach that isn't reflected here is what Redis does. Specifically: Stores and updates on the working set are held in memory, and logged behind to persistent storage (with explicit sync points for operations you need to be sure landed on disk). Possibly we can't do this if our working set is expected to be too big for memory, but the strategy seems like a good possible fit for our usage patterns.

Reply:
> If blocks are never mutated and are append only do we even need a locking mechanism? We're not going to have a situation of conflicting reads/writes

Blocks aren't modified, indexes and various caches presumably will. (e.g. if we persist pointers on disk to "latest" things, as well as the WAL. 

I guess what I'm arguing for is more that tendermint should rely on the database engine's concurrency control and shouldn't need to add an additional mutex around the database. 

---- 

Redis is delightfully simple, and I think there is merit to their approach, from a high level, but I think a big part of this whole project is to reduce the current level of complexity in the relationship between the consensus engine (tendermint) and the storage engine while also reducing (somewhat) the porous boundary between these components. It feels like doing "the redis thing" would be something of a premature optimization at this point.